telemetry/src/metrics.rs

```diff
@@ -0,0 +1,174 @@
+            peers: val.connected_peers.get(),
+            blocks: val.block_height.get(),
+            txs: val.txs.with_label_values(&["total"]).get(),
+            uptime: Uptime(Duration::from_millis(val.uptime_since_genesis_ms.get())),
+        }
+    }
+}
+
+/// A strict superset of [`Status`].
+#[derive(Debug)]
+pub struct Metrics {
+    /// Transactions in the last committed block
+    pub txs: IntCounterVec,
+    /// Current block height
+    pub block_height: IntCounter,
+    /// Total number of currently connected peers
+    pub connected_peers: GenericGauge<AtomicU64>,
+    /// Uptime of the network, starting from commit of the genesis block
+    pub uptime_since_genesis_ms: GenericGauge<AtomicU64>,
+    /// Number of domains.
+    pub domains: GenericGauge<AtomicU64>,
+    /// Total number of users per domain
+    pub accounts: GenericGaugeVec<AtomicU64>,
+    /// Transaction amounts.
+    pub tx_amounts: Histogram,
+    /// Queries handled by this peer
+    pub isi: IntCounterVec,
+    /// Query handle time Histogram
+    pub isi_times: HistogramVec,
+    // Internal use only.
+    registry: Registry,
+}
+
+impl Default for Metrics {
+    // The constructors either always fail, or never.
+    #[allow(clippy::expect_used)]
+    fn default() -> Self {
+        let txs = IntCounterVec::new(Opts::new("txs", "Transactions committed"), &["type"])
+            .expect("Infallible");
+        let isi = IntCounterVec::new(
+            Opts::new("isi", "Iroha special instructions handled by this peer"),
+            &["type", "success_status"],
+        )
+        .expect("Infallible");
+        let isi_times = HistogramVec::new(
+            HistogramOpts::new("isi_times", "Time to handle isi in this peer"),
+            &["type"],
+        )
+        .expect("Infallible");
+        let tx_amounts = Histogram::with_opts(HistogramOpts::new(
+            "tx_amount",
+            "average amount involved in a transaction on this peer",
+        ))
+        .expect("Infallible");
+        let block_height =
+            IntCounter::new("block_height", "Current block height").expect("Infallible");
+        let connected_peers = GenericGauge::new(
+            "connected_peers",
+            "Total number of currently connected peers",
+        )
+        .expect("Infallible");
+        let uptime_since_genesis_ms = GenericGauge::new(
+            "uptime_since_genesis_ms",
+            "Network up-time, from creation of the genesis block",
+        )
+        .expect("Infallible");
+        let domains = GenericGauge::new("domains", "Total number of domains").expect("Infallible");
+        let accounts = GenericGaugeVec::new(
+            Opts::new("accounts", "User accounts registered at this time"),
+            &["domain"],
+        )
+        .expect("Infallible");
+        let registry = Registry::new();
+
+        macro_rules! register {
+			($metric:expr)=> {
+				registry.register(Box::new($metric.clone())).expect("Infallible");
```

Comment:
should registry share references to metrics being tracked or just a clone?

Reply:
Metrics internally use `Arc`. 
See their official examples: 
https://github.com/tikv/rust-prometheus/blob/e3e2ffdd566b1c0ce19e2be8a1dc849f8f88c4ac/examples/example_custom_registry.rs#L47